src/placeos-driver/interface/sensor.cr

```diff
@@ -0,0 +1,137 @@
+    Switch      # On or Off - water leak, door open etc
+    Level       # Percentage + raw value + technical human readable (battery, fuel, water tank, oxygen, Co2)
+    Flow        # https://en.wikipedia.org/wiki/Cubic_metre_per_second (gas or liquid)
+    Counter
+    Acceleration # MetrePerSecondSquared
+    Speed        # MetrePerSecond
+    # Gyroscope: roll, pitch, yaw - Angle
+    Roll
+    Pitch
+    Yaw
+    Compass              # direction (magnetic field)
+    Current              # https://en.wikipedia.org/wiki/Ampere
+    Voltage              # https://en.wikipedia.org/wiki/Volt
+    ElectricalResistance # https://en.wikipedia.org/wiki/Ohm
+    Radiation            # https://en.wikipedia.org/wiki/Sievert
+    Distance
+  end
+
+  # Using SI units and SI derived units
+  enum Unit
+    Celsius               # Temperature - celsius over Kelvin to avoid some conversions
+    Percentage            # Humidity, Level
+    Lux                   # Illuminance
+    Pascal                # Pressure
+    Boolean               # Trigger, Switch
+    Integer               # Counter
+    CubicMetrePerSecond   # Liquid or gas flow rate
+    MetrePerSecondSquared # Acceleration
+    MetrePerSecond        # Speed
+    Angle                 # Compass, Accel, Gyro
+    Ampere                # Current
+    Volt                  # Voltage
+    Ohm                   # ElectricalResistance
+    Sievert               # Radiation
+    Metre                 # Distance
+  end
+
+  # return the specified sensor details
+  abstract def sensor(unique_id : String) : Detail?
+
+  # return an array of sensor details
+  # zone_id can be ignored if location is unknown by the sensor provider
+  # mac_address can be used to grab data from a single device (basic grouping)
+  abstract def sensors(type : String? = nil, mac_address : String? = nil, zone_id : String? = nil) : Array(Detail)
+
+  abstract class Location
+    include JSON::Serializable
+
+    use_json_discriminator "type", {
+      "geo" => GeoLocation,
+      "map" => MapLocation,
+    }
+  end
+
+  class GeoLocation < Location
+    getter type : String = "geo"
+
+    def initialize(@lat, @lon)
+    end
+
+    property lat : Float64
+    property lon : Float64
+
+    def s2_cell_id(s2_level : Int32 = 21)
+      S2Cells::LatLon.new(lat, lon).to_token(s2_level)
+    end
+  end
+
+  class MapLocation < Location
+    getter type : String = "map"
+
+    def initialize(@x, @y)
+    end
+
+    # this is expected to be the raw unadjusted values
+    property x : Float64
+    property y : Float64
+  end
+
+  abstract class Detail
+    include JSON::Serializable
+
+    def initialize(@type, @unit, @value, @unix_ms, @mac, @id, @name, @raw, @loc)
+    end
+
+    property type : SensorType
+    property unit : Unit
+
+    property value : Float64
+    property unix_ms : Int64
+
+    # the unique id can be optional if the mac address represents a single value
+    property mac : String
+    property id : String?
```

Comment:
Do both of these need to be kept? Why not just a single id to remove potential confusion in usage.

Reply:
yeah as a single device might have multiple sensors
So MAC allows you to group them and id to uniquely identify the individual sensor.

Although we possibly just combine the two to generate a unique id - basically this is the datapoints primary key